File: BOJ/단계별 문제/11. 브루트 포스/1. 블랙잭/dfs.cpp
#include <iostream>
#include <vector>
using namespace std;

int N, M, ans = 0;
vector<int> nums;

void f(int depth, int pos, int sum) {
    if (depth == 3) {
        ans = sum <= M? max(ans, sum): ans;
        return;
    }

    for (int i=pos; i<N; i++)
        f(depth+1, i+1, sum+nums[i]);
}

int main() {
    cin >> N >> M;

    for (int i=0; i<N; i++) {
        int tmp;
        cin >> tmp;
        nums.push_back(tmp);
    }

    f(0, 0, 0);
    cout << ans;

    return 0;
}

File: BOJ/단계별 문제/11. 브루트 포스/1. 블랙잭/dfs.py
N, M = map(int, input().split())
nums = list(map(int, input().split()))
ans = 0

def f(depth, pos, _sum):
    global N, M, nums, ans

    if depth == 3:
        ans = max(ans, _sum) if _sum <= M else ans
        return

    for i in range(pos, N):
        f(depth+1, i+1, _sum+nums[i])

f(0, 0, 0)
print(ans)
